Source/DafnyTestGeneration/Auditor.cs

```diff
@@ -0,0 +1,373 @@
+  /// the type of the parameter is fully supported by test generation. Return false, if this is not the case.
+  /// </summary>
+  private bool CheckInputTypesAreSupported(Program program) {
+    var result = true;
+    foreach (MemberDecl declaration in Utils.AllMemberDeclarationsWithAttribute(program.DefaultModule, TestGenerationOptions.TestEntryAttribute)) {
+      if (declaration.EnclosingClass is TraitDecl or ArrayClassDecl or IteratorDecl) {
+        diagnostics.Add(new DafnyDiagnostic(UnsupportedInputTypeError, declaration.Tok,
+          $"Test Generation does not support trait, array, or iterator types as receivers of " +
+          $"{{:{TestGenerationOptions.TestEntryAttribute}}}-annotated methods.\n" +
+          $"Consider writing a wrapper method that creates a receiver and passes on the arguments to it",
+          MessageSource.TestGenerationAuditor, ErrorLevel.Error, new List<DafnyRelatedInformation>()));
+        result = false;
+      } else if (declaration.EnclosingClass is ClassDecl) {
+        diagnostics.Add(new DafnyDiagnostic(NotFullySupportedInputTypeWarning, declaration.Tok,
+          $"Test Generation does not fully support class types as receivers of " +
+          $"{{:{TestGenerationOptions.TestEntryAttribute}}}-annotated methods.\n" +
+          $"Consider writing a wrapper method that creates a receiver and passes on the arguments to it",
+          MessageSource.TestGenerationAuditor, ErrorLevel.Warning, new List<DafnyRelatedInformation>()));
+        result = false;
+      }
+      if (declaration is Method method) {
+        foreach (var formal in method.Ins) {
+          if (!TypeIsSupported(formal.Type, method.Name)) {
+            result = false; // No short-circuiting here to emit all possible errors/warnings
+          }
+        }
+      } else if (declaration is Function function) {
+        foreach (var formal in function.Formals) {
+          if (!TypeIsSupported(formal.Type, function.Name)) {
+            result = false; // No short-circuiting here to emit all possible errors/warnings
+          }
+        }
+      }
+    }
+    return result;
+  }
+
+  /// <summary>
+  /// Recursively traverse a given type and emit warning/errors if this type cannot be fully reconstructed
+  /// from the counterexample by test generation, i.e. no parameters of the {:testEntry}-annotated method/function
+  /// should have such a type. Return true, if the type is fully supported. 
+  /// </summary>
+  private bool TypeIsSupported(Type type, string testEntry) {
+    if (typesConsidered.Contains(type)) {
+      return true;
+    }
+    typesConsidered.Add(type);
+    var isSupported = false;
+    while (type is InferredTypeProxy inferred) {
+      type = inferred.T;
+    }
+    if (type is UserDefinedType userDefinedType) {
+      var genericMessage =
+        $"Consider disallowing values of type {userDefinedType} to be part of the input to " +
```

Comment:
The word "disallowing" seems weird here. Maybe use terminology similar to what you use elsewhere about writing a wrapper to generate types that can't be constructed automatically out of types that can?

Reply:
I have paraphrased this to "Consider modeling values of type TYPE with a datatype and passing them as input to the {:testEntry} annotated method". I agree that something along these lines should be better than "disallowing" since it tells the user what to do to prevent the error.